### ISSUE TITLE
Long term GRU time series prediction issue 

### ISSUE THREAD

**Issue by ArashAskary** [rstudio/keras#517]
I am trying to build a timeseries forecasting based on deep learning using a GRU. I have managed to get a single prediction. For reference, if anyone would like to replicate my process, here is my data - 
https://drive.google.com/file/d/1nTQjkMuYFVaYFP4mTtjTZkEXCLwjmGip/view?usp=sharing

Here is what my data looks like. I'm attempting to forecast the 'Total.Sum' feature (note that both variables have been transformed) - 

    str(df)

    'data.frame':	268 obs. of  5 variables:
     $ Total.Sum: num  0.0159 0.0233 0.0234 0.024 0.0157 ...
     $ weekday  : num  0.25 0.5 0.75 1 0 0.25 0.5 0.75 1 0 ...

I have 256 training examples. 
Pre-processing the data for the model - 
    
    # Splitting into training and testing set
    df <- as.matrix(df)
    x_train <- array_reshape(df[1:256,], dim = c(256,2,1)) 
    y_train <- df[,1] 
    y_train <- array_reshape(df, dim = c(256, 1)) 

And here's my GRU model - 

    model <- keras_model_sequential()
    model %>% 
      layer_gru(units = 32, batch_input_shape = c(256, 2, 1), return_sequences = T, stateful = T, activation = "relu") %>%
      layer_gru(units = 64, return_sequences = T, stateful = T, activation = "relu") %>%
      layer_gru(units = 128, return_sequences = T, stateful = T, activation = "relu") %>%
      layer_gru(units = 32, return_sequences = F, stateful = T, activation = "relu") %>%
      layer_dense(units = 1) %>%
      layer_activation(activation = "linear")

    model %>% compile(loss = "mse", optimizer = "adam")

Finally, here's my training loop - 
    
    for (i in 1:1000) {
      model %>% fit(x_train, y_train, batch_size = 256, shuffle = FALSE, epochs = 1)
      model %>% reset_states()
    }

My question is, how can I feed in the model's predictions in order to acquire a longer-term forecast? I have been attempting to understand how this was achieved from this post - https://stackoverflow.com/questions/38714959/understanding-keras-lstms/50235563#50235563, but have found it difficult to comprehend, especially when attempting to translate from python to R. From what I understand, we need to predict the next step in Y, and somehow loop that back into our model... 

From what I gather, I need to build a new model that can accept a tensor of variable lengths (i.e list(NULL, 2, 1))?

This is how far I've gotten with the forecast -
```

model %>% reset_states()
predictions <- model %>% predict(x_train, batch_size = 256)  # This creates states
future <- c()
current_step <- predictions [c(256)] # Selecting the last result

for (i in seq(60)){
  current_step <- array_reshape(current_step, dim = c(1,2,1))
  current_step <- model.simp %>% predict(current_step, batch_size = 1)
  future[i] <- current_step
}
```

Any and all help is much appreciated! 

  [1]: https://drive.google.com/file/d/1nTQjkMuYFVaYFP4mTtjTZkEXCLwjmGip/view?usp=sharing

**Comment by skeydan** [comment]
Hi,

for the problem of forecasting multiple timesteps, @kevinykuo has a great explanation of how to do this here: 

https://github.com/kevinykuo/user2018-deeplearning/blob/master/notebooks/4_sunspots.Rmd

Basically  your options would be method 3 and 4. 

(3) is easier, but it presupposes it's ok for you to output as many predictions as you're considering in the RNN _timesteps_ dimension.

For example, say your predictions are based on 12 months, then you can predict 12 months.
This way is explained in detail in 

https://blogs.rstudio.com/tensorflow/posts/2018-06-25-sunspots-lstm/

(4) gives you total freedom, but it is a bit more involved. It is based on the encoder-decoder architecture explained in e.g..

https://blog.keras.io/a-ten-minute-introduction-to-sequence-to-sequence-learning-in-keras.html
